lib/pbench/server/api/resources/upload_api.py

```diff
@@ -196,11 +198,34 @@ def put(self, filename: str):
 
                     ofp.write(chunk)
                     hash_md5.update(chunk)
-            except Exception:
+            except OSError as exc:
+                if exc.errno == errno.ENOSPC:
+                    self.logger.error(
+                        "Not enough space on volume, {}, for upload:"
+                        " user = {}, ctrl = {!a}, file = {!a}",
+                        path,
+                        username,
+                        controller,
+                        filename,
+                    )
+                    abort(HTTPStatus.INSUFFICIENT_STORAGE)
+                else:
+                    msg = "Unexpected error encountered during file upload"
+                    self.logger.error(
+                        "{}, {}, for user = {}, ctrl = {!a}, file = {!a}",
+                        msg,
+                        exc,
+                        username,
+                        controller,
+                        filename,
+                    )
+                    abort(HTTPStatus.INTERNAL_SERVER_ERROR, message="INTERNAL ERROR")
+            except Exception as exc:
                 msg = "Unexpected error encountered during file upload"
-                self.logger.exception(
-                    "{} for user = {}, ctrl = {!a}, file = {!a}",
+                self.logger.error(
+                    "{}, {}, for user = {}, ctrl = {!a}, file = {!a}",
                     msg,
+                    exc,
                     username,
                     controller,
                     filename,
```

Comment:
Having to repeat lines 224-233 at 213-222 is disappointing.  We should consider implementing a utility function like the following:
```
def errno_exception(errno_val: int):
    exc = sys.exc_info()[0]
    if exc is OSError and sys.exc_info()[1].errno == errno_val:
        return exc
    # Return any exception class which doesn't match.                                                                                                                                          
    return ValueError if exc is not ValueError else TypeError
```
We could then use this at line 201 to catch only the specific `OSError` that we're interested in:
```
            except errno_exception(errno.ENOSPC) as exc:
```
And any other `OSError`'s would fall through to the catch-call case.

This would presumably be useful in other places where we catch `OSError`....

Reply:
Tricky.

I'll have to noodle on that a bit.

Tracked in #2322.